cmd/snap/cmd_run.go

```diff
@@ -133,15 +135,32 @@ func createOrUpdateUserDataSymlink(info *snap.Info, usr *user.User) error {
 			// TRANSLATORS: %v the error message
 			return fmt.Errorf(i18n.G("cannot read symlink: %v"), err)
 		}
+
 		if currentSymlinkValue == wantedSymlinkValue {
 			break
 		}
+
 		if !os.IsNotExist(err) {
-			_ = os.Remove(currentActiveSymlink)
+			// We may be racing with other instances of snap-run that try to do the same thing
+			// If the symlink is already removed then we can ignore this error and try again.
+			err = os.Remove(currentActiveSymlink)
+			if os.IsNotExist(err) {
```

Comment:
I think the logic is correct, although I agree that line 143 is very confusing; when we arrive at 143 the file either exists (err == nil) or not, i.e. IsNotExist(err) is true; on all other Readlink errors we abort on the error check after Readlink above. I agree continuing was a bad idea since it's exhausting an iteration of retry for no good reason.

Nonetheless, I've updated the logic, removed Time.Sleep delays and added a new case in the spread test to check that incorrect version link is removed and recreated.